Can't import TS packages since v20.1.0

### Current Behavior

I'm using all the latest versions of all packages as of now, nx started to fail to build my projects, because when I try to import one of my own packages - it can't resolve them since v20.1.0. I tried reinstalling and doing full git clean, nothing helped. In v20.0.13 all works fine.

### Expected Behavior

All builds like in v20.0.13

### GitHub Repo

_No response_

### Steps to Reproduce

I have paths like so in my tsconfig: `"@nx/math": ["packages/math/src/index.ts"]`
And I'm using `@nx/js:tsc` to build projects, like so:
```json
"build": {
  "executor": "@nx/js:tsc",
  "outputs": ["{options.outputPath}"],
  "options": {
    "outputPath": "dist/packages/math",
    "main": "packages/math/src/index.ts",
    "tsConfig": "packages/math/tsconfig.library.json",
    "assets": ["packages/math/*.md"],
    "outputFileName": "src/index.js"
  }
},
```

### Nx Report

```shell
Node           : 20.16.0
OS             : win32-x64
Native Target  : x86_64-windows
npm            : 10.8.1

nx                 : 20.1.0
@nx/js             : 20.1.0
@nx/jest           : 20.1.0
@nx/eslint         : 20.1.0
@nx/workspace      : 20.1.0
@nx/devkit         : 20.1.0
@nx/eslint-plugin  : 20.1.0
@nx/plugin         : 20.1.0
@nx/vite           : 20.1.0
typescript         : 5.6.3
---------------------------------------
Local workspace plugins:
         @nx/nx-stuff
```

### Failure Logs

```shell
/Users/maxim.mazurok/transactions/packages/rates/src/lib/rates.ts:3:43 - error TS7016: Could not find a declaration file for module '@nx/math'. '/Users/maxim.mazurok/transactions/dist/packages/math/src/index.js' implicitly has an 'any' type.
        Try `npm i --save-dev @types/nx__math` if it exists or add a new declaration (.d.ts) file containing `declare module '@nx/math';`

      3 import { mathjsNumber, parseNumber } from "@nx/math";
                                                  ~~~~~~~~~~
```

### Package Manager Version

_No response_

### Operating System

- [ ] macOS
- [ ] Linux
- [x] Windows
- [ ] Other (Please specify)

### Additional Information

Potentially related: https://github.com/nrwl/nx/pull/28446 ( @leosvelperez )

--- a/packages/js/src/utils/package-json/update-package-json.spec.ts
+++ b/packages/js/src/utils/package-json/update-package-json.spec.ts
@@ -405,6 +405,38 @@ describe('getUpdatedPackageJsonContent', () => {
       },
     });
   });
+
+  it('should handle outputFileName correctly', () => {
+    expect(
+      getUpdatedPackageJsonContent(
+        {
+          name: 'test',
+          version: '0.0.1',
+        },
+        {
+          main: 'proj/src/index.ts',
+          outputPath: 'dist/proj',
+          projectRoot: 'proj',
+          format: ['cjs'],
+          generateExportsField: true,
+          outputFileName: 'src/index.js',
+        }
+      )
+    ).toEqual({
+      name: 'test',
+      main: './src/index.js',
+      types: './src/index.d.ts',
+      version: '0.0.1',
+      type: 'commonjs',
+      exports: {
+        '.': {
+          default: './src/index.js',
+          types: './src/index.d.ts',
+        },
+        './package.json': './package.json',
+      },
+    });
+  });
 });
 
 describe('updatePackageJson', () => {

--- a/packages/js/src/utils/package-json/update-package-json.ts
+++ b/packages/js/src/utils/package-json/update-package-json.ts
@@ -242,7 +242,7 @@ export function getExports(
 ): Exports {
   const outputDir = getOutputDir(options);
   const mainFile = options.outputFileName
-    ? options.outputFileName.replace(/\.[tj]s$/, '')
+    ? basename(options.outputFileName).replace(/\.[tj]s$/, '')
     : basename(options.main).replace(/\.[tj]s$/, '');
   const exports: Exports = {
     '.': outputDir + mainFile + options.fileExt,
@@ -377,9 +377,9 @@ export function getOutputDir(
     : options.outputPath;
   const relativeOutputPath = relative(packageJsonDir, options.outputPath);
   const relativeMainDir = options.outputFileName
-    ? ''
+    ? dirname(options.outputFileName)
     : relative(options.rootDir ?? options.projectRoot, dirname(options.main));
-  const outputDir = join(relativeOutputPath, relativeMainDir);
+  const outputDir = joinPathFragments(relativeOutputPath, relativeMainDir);
 
   return outputDir === '.' ? `./` : `./${outputDir}/`;
 }

